New working method (Fix #11)

ImJeanRobert opened the pull request:
This new method changes the original player of the website with the embeded one, which show no ads.
It also remodel it to make it look like the original one, so you're not lost ;)

This fixes #11

ImJeanRobert committed:
Added new method with embed

ImJeanRobert committed:
Deleted old file

ImJeanRobert committed:
Delete options.html

ImJeanRobert committed:
Delete options.js

ImJeanRobert committed:
Delete user-agent.txt

ImJeanRobert committed:
Update README.md

ImJeanRobert committed:
Re format fucked up zones

ImJeanRobert committed:
Fix streamer main page

It fixes the streamer main page (the weird page with the stream in the background and access to videos, clips and so forth)

ImJeanRobert committed:
Fix mature content popup & Add AudioCompressor

- Fix mature content popup
- Add AudioCompressor (thanks to @rmanky)

ImJeanRobert committed:
Fixed typo

WHCTHBD commented:
Mainfest fix attached.

`	"permissions": [
		"*://*.twitch.tv/*",
		"*://*.ttvnw.net/*",
		"activeTab",
		"storage",
		"webRequest",
		"webRequestBlocking"
	]`

[manifest.zip](https://github.com/odensc/ttv-ublock/files/5578142/manifest.zip)

ImJeanRobert commented:
I checked that on Edge, Chrome and Firefox and the current manifest has no problem.

But there is one bug : The audio don't play on Firefox.
If someone can look into this...

ImJeanRobert committed:
Fix Theater button positioning on Firefox

WHCTHBD commented:
> I checked that on Edge, Chrome and Firefox and the current manifest has no problem.
> 
> But there is one bug : The audio don't play on Firefox.
> If someone can look into this...

That's bizarre re: Manifest. What version of Chrome are you on? 86.0.4240.198 here.

ImJeanRobert committed:
Fix last update for Chrome (firefox & chrome work)

Last update broke chrome.
Now it works perfectly fine on chrome and firefox and the theater button is on the right position on both browsers.

@r3nderer you can implement this in your script if you want.

ImJeanRobert commented:
It worked on chrome 86.0.4240.198 and 87.0.4280.66 (latest)

rmanky commented:
Not sure what's going on with Firefox, I've tried a variety of things to no avail...

Will update if I find anything

Edit: I believe the issue is this `The HTMLMediaElement passed to createMediaElementSource has a cross-origin resource, the node will output silence.`, still trying to fix

ImJeanRobert committed:
Fix audio in Firefox

The AudioContext class used was the class from the original website context and not the iframe's one so it didn't work.

@rmanky

ImJeanRobert commented:
Okay, I fixed it, you were true, I failed the implementation and it was causing something like a cross origin resource problem.
@rmanky

ImJeanRobert commented:
@odensc Can you please merge or is there something you think we should update ?

asianboy808 commented:
I'm just looking at this code as an outsider. I think it's great that we came up with a working solution, and we should definitely deploy it to those that are consuming ttv-ublock early. But I do have a few short term concerns and long term concerns as the code evolves.

* Before this PR, the code was very simple and only hooked into web request headers, only needing about 25 lines of code. This by itself meant it was ultra-maintanable to short term changes.
* This PR introduces a large change where we are replacing the entire video player, which is simple enough to do now, but we need to be ready to upgrade this code as Twitch fixes their stuff. If you are not available to do so, others have to come in to learn to understand it, and the more we tack on, the more difficult it'll be to handle the code. What I suggest here is that we spend some time to make it cleaner and compartmentalize the code so it's more digestible. 
* It looks like we make a request to replace the player in this code, does that introduce any negative drawbacks? For example, alternative player for twitch adds a few seconds of delay to a stream. For some people such as myself, this is a huge negative part that made me look for other solutions. It's important to outline the caveats in this change. 
* I can't speak for oden, but if I owned this project, I would like to see the code look cleaner in that we are consistent with aspects of the codebase that existed beforehand. For instance, we completely deleted the files that were previously existing, and replaced the project. The new code uses spaces instead of tabs like the previous version. The new code uses `var` instead of `let` and `const`.  I'm sorry, but as a professional developer myself, these tiny issues stack up and just make it a bit harder to work on the codebase. I want to provide feedback after this comment to try to make it look a bit nicer. 
* We should also spend our knowledge diving into better ways to fight twitch's system, because these rudimentary solutions that replace the window fetch, replace header user agents, or replace the player are clearly not enough. Developers can patch these quickly. I've personally been investigating things that developers may not be able to patch up that quickly, or at least spend a few sprint cycles investigating. For example, hijacking parts of the GraphQL queries that twitch makes, and seeing if we can straight up send fake queries to cancel the ads, or injecting stuff into parts of the React `withVideoPlayerContext` to see if we can replace some of the functions they have built in. I haven't actually gotten anything working yet, but my thought is targeting code infrastructure they have put in place will require their developers to focus on more obfuscation as well as even renaming parts of their codebase. This takes a lot of time and money on their end for the people working against us.

asianboy808 commented on options.html:
```diff
@@ -1,23 +0,0 @@
-<!DOCTYPE html>
```
Why are we deleting this part of the code? Shouldn't we keep the shell at least, in case we want to add stuff to it later on?

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+var lastStreamer, existingIframe;
+
+window.addEventListener("message", (event) => {
```
How often does this occur? Will this script have the potential to lag out? For `embAdbQuality`, is that something that twitch has set that we are trying to inject something to so twitch reads a different value? Or is it just something we are storing to refer to later, in which case wouldn't a variable we define be better? 

I do see down below we use this `localStorage` but I can't help but think that we can optimize it just by using a variable.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+
+window.addEventListener("message", (event) => {
+  if (event.data.eventName == "UPDATE_STATE" && event.data.params.quality && !document.hidden)
+    if (/^((?:160|360|480|720|900|960|1080)p(?:30|48|50|60)|chunked)$/.test(event.data.params.quality))
```
nit: Can we create a constant for this regex?

asianboy808 replied:
And also another variable for `event.data.params.quality`, or deconstruct the usage of them:

```js
const { params } = event.data;
const { quality } = params;
```

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+var lastStreamer, existingIframe;
```
nit: let's be consistent with let vs var. Because we're likely targeting users who have Chrome/Firefox, (fuck IE lol) just use `let`. `var` has too many inconsistencies with variable scope.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+  var container = document.querySelector(".video-player .tw-absolute");
+
+  //Optimization: We do not check everything when numbers like viewers or anything updates.
+  if (mutations.length === 1 && (mutations[0].target.classList.contains("tw-animated-number--monospaced") || mutations[0].target.classList.contains("tw-animated-number")))
```
nit: long line of code, I personally follow airbnb style guide for this.

```suggestion
  if (
    mutations.length === 1
    && (mutations[0].target.classList.contains("tw-animated-number--monospaced")
      || mutations[0].target.classList.contains("tw-animated-number")
    )
 )
```
or something like this

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+const untouchedPages = [(link) => link.endsWith("/directory"), (link) => link.includes("/clip/")];
+
+var observer = new MutationObserver(function (mutations, observer) {
+  var container = document.querySelector(".video-player .tw-absolute");
```
Maybe we can create constants for these in case twitch ever decides to change these to fuck with us. For example, it isn't that hard to append a key/randomly generated ID to classes if you're using a build system like webpack which they likely are doing with their front end code.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+    
+  //If there is no container, then the page is not fully loaded
+  if (!container)
+    return;
```
This one is up to oden but I personally like adding `{}` all the time, following the airbnb style guide.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+  //Thses pages should not be modified because it breaks them
+  if (untouchedPages.find(value => value(window.location.pathname) || (isOnMainPage && !lastStreamer))) {
+    //We show the initial interface and hide our embeded stream
+    existingIframe?.remove();
```
Is the optional chaining available in both Chrome and Firefox already? I wonder how many people keep their browsers up to date to support this. Maybe we should stick with `existingIframe && existingIframe.remove()` instead.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+    existingIframe?.remove();
+    existingIframe = undefined;
+
+    for (let el of container.children)
+      el.hidden = false;
+    return;
+  } else if(isOnMainPage)//If the user is on the main page, we do not modify anything (and we do not revert our modifications (because of the little player at the bottom of the page))
+    return;
```
nit: somewhere along the line the spacing got a bit fucked up and it doesn't look very nice

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+
+  //We save the latest stream quality choosen by the user so we use it there to tell the embeded iframe what quality to use
+  let quality = localStorage["embAdbQuality"] || "chunked";
+  let iframeUrl = `https://player.twitch.tv/?channel=${streamerName}&muted=false&parent=twitch.tv&quality=${quality}`;
```
Does this make the network request once everything is loaded, or as everything is loading? One concern is that twitch may be able to just block requests like this from happening depending on how or when we request it.

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+});
+
+//Code modifying the iframe
+function loadIframe(iframe) {
```
This is a giant function. Maybe we can split it off into pieces (for example wherever we have comments explaining a chunk of code, we split that into its own function.

asianboy808 replied:
or class

asianboy808 commented on scripts/mainsite.js:
```diff
@@ -0,0 +1,213 @@
+    let config = document.querySelector('[data-a-target="player-settings-button"]');
+    //The interface is not made in the same way using firefox or chrome
+    //Firefox way
+    if(fullscreenButton.parentElement.children.length > 3)
+      config.parentElement.parentElement.insertBefore(
+        theaterButton.parentElement,
+        config.parentElement.nextSibling
+      );
+    else//Chrome way
+      fullscreenButton.parentElement.parentElement.insertBefore(
+        theaterButton.parentElement,
+        fullscreenButton.parentElement
+      );
```
```suggestion
const isFirefox = fullscreenButton.parentElement.children.length > 3; // could also use browser detection here

if (isFirefox) {
...
} else {
...
} 
```

This way conveys through code the idea instead of through the comments. In essence, it's cleaner.

ImJeanRobert replied:
Thank you for all your suggestions, I'll look into this later since I don't have the time now (or if someone can do so before, it could be cool).

For the message handling thing, we are storing the latest resolution choosen by the user in the localstorage to reuse it when an iframe is append (we set the default resolution to the last used one).

We are loading the iframe only when the url should change, so it is not so often.

There is already some optimizations in place to avoid consuming too much cpu but maybe could be better.

The regex can be changed to a list, you're true

odensc commented:
Thanks for all the effort you've put into this! I echo some of @asianboy808's concerns, mostly that with how deep this integrates into Twitch's site, it has a high potential to break whenever they deploy a new release, and cause users of the extension to have issues watching streams.

I plan on implementing the simpler method of faking an embed player (eg https://github.com/odensc/ttv-ublock/issues/11#issuecomment-731692364) via network requests, and if (once) they patch that, I'll consider other solutions such as this.

odensc commented:
Closing for now due to the new extension. Again thanks for all the work.
